File: Hard/data_stream_as_disjoint_intervals/data_stream_as_disjoint_intervals.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

class SummaryRanges {
    public:
        vector<vector<int>> intervals;

        SummaryRanges() {

        }

    void addNum(int value) {
        vector<vector<int>> newIntervals;
        int start = value;
        int end = value;

        for (vector<int>& interval : intervals) {

            // Current interval comes before the new value
            if (interval[1] < start - 1) {
                newIntervals.push_back(interval);
            }
            // Current interval comes after the new value
            else if (interval[0] > end + 1) {
                newIntervals.push_back({start, end});
                start = interval[0];
                end = interval[1];
            }

            // Intervals overlap or touch
            else {
                start = min(start, interval[0]);
                end = max(end, interval[1]);
            }
        }

        newIntervals.push_back({start, end});
        intervals = newIntervals;
    }

    vector<vector<int>> getIntervals() {
        return intervals;
    }
};

// Example usage:
int main() {
    SummaryRanges summaryRanges;
    summaryRanges.addNum(1);
    cout << "Intervals after adding 1: ";
    for (const auto& interval : summaryRanges.getIntervals()) {
        cout << "[" << interval[0] << ", " << interval[1] << "] ";
    }
    cout << endl;
    summaryRanges.addNum(3);
    cout << "Intervals after adding 3: ";
    for (const auto& interval : summaryRanges.getIntervals()) {
        cout << "[" << interval[0] << ", " << interval[1] << "] ";
    }
    cout << endl;
    summaryRanges.addNum(7);
    cout << "Intervals after adding 7: ";
    for (const auto& interval : summaryRanges.getIntervals()) {
        cout << "[" << interval[0] << ", " << interval[1] << "] ";
    }
    cout << endl;
    summaryRanges.addNum(2);
    cout << "Intervals after adding 2: ";
    for (const auto& interval : summaryRanges.getIntervals()) {
        cout << "[" << interval[0] << ", " << interval[1] << "] ";
    }
    cout << endl;   

    summaryRanges.addNum(6);
    cout << "Intervals after adding 6: ";
    for (const auto& interval : summaryRanges.getIntervals()) {     
        cout << "[" << interval[0] << ", " << interval[1] << "] ";
    }
    cout << endl;   

    return 0;
}

File: Hard/data_stream_as_disjoint_intervals/data_stream_as_disjoint_intervals.py
from typing import List

class SummaryRanges:
    def __init__(self):
        self.intervals = []

    def addNum(self, value: int) -> None:
        new_intervals = []
        start = value
        end = value
        for interval in self.intervals:
            if interval[1] < start - 1: #Current interval comes before the new value
                new_intervals.append(interval)
            elif interval[0] > end + 1: #Current interval comes after the new value
                new_intervals.append([start, end])
                start = interval[0]
                end = interval[1]
            else: #Current interval overlaps with the new value
                start = min(start, interval[0])
                end = max(end, interval[1])
        new_intervals.append([start, end])
        self.intervals = new_intervals

    def getIntervals(self) -> List[List[int]]:
        return self.intervals

if __name__ == "__main__":
    print("Testing SummaryRanges class: ")
    summary_ranges = SummaryRanges()
    summary_ranges.addNum(1)
    print("After adding 1, intervals are:", summary_ranges.getIntervals()) #Output: [[1, 1]]
    summary_ranges.addNum(3)
    print("After adding 3, intervals are:", summary_ranges.getIntervals()) #Output: [[1, 1], [3, 3]]
    summary_ranges.addNum(7)
    print("After adding 7, intervals are:", summary_ranges.getIntervals()) #Output: [[1, 1], [3, 3], [7, 7]]
    summary_ranges.addNum(2)
    print("After adding 2, intervals are:", summary_ranges.getIntervals()) #Output: [[1, 3], [7, 7]]
    summary_ranges.addNum(6)
    print("After adding 6, intervals are:", summary_ranges.getIntervals ()) #Output: [[1, 3], [6, 7]]